opal/mca/btl/ofi/btl_ofi_providers.h

```diff
@@ -0,0 +1,98 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil -*- */
+/*
+ * Copyright (c) 2018      Intel, Inc, All rights reserved
+ * Copyright (c) 2018      Los Alamos National Security, LLC. All rights
+ *                         reserved.
+ *
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "btl_ofi.h"
+
+#define MCA_BTL_OFI_REQUIRED_CAPS       (FI_RMA | FI_ATOMIC)
+#define MCA_BTL_OFI_REQUESTED_MR_MODE   (FI_MR_UNSPEC)
+
+/* handle provider specific hints */
+static void provider_hints_handler(struct fi_info *hints);
+
+
+/* info validation after fi_getinfo() */
+static int validate_info(struct fi_info *info);
+static int socks_validate_info(struct fi_info *info);
+static int psm2_validate_info(struct fi_info *info);
+static int gni_validate_info(struct fi_info *info);
+
+
+/* functions below */
+
+static int socks_validate_info(struct fi_info *info)
+{
+    /* placeholder */
+    return OPAL_SUCCESS;
+}
+
+static int psm2_validate_info(struct fi_info *info)
+{
+    /* placeholder */
+    return OPAL_SUCCESS;
+}
+
+static int gni_validate_info(struct fi_info *info)
+{
+    /* placeholder */
+    return OPAL_SUCCESS;
+}
+
+/* validate information returned from fi_getinfo().
+ * return OPAL_ERROR if we dont have what we need. */
+static int validate_info(struct fi_info *info)
+{
+    int mr_mode;
+    char *prov_name = info->fabric_attr->prov_name;
+
+    /* we need exactly all the required bits */
+    if ((info->caps & MCA_BTL_OFI_REQUIRED_CAPS) != MCA_BTL_OFI_REQUIRED_CAPS)
+        return OPAL_ERROR;
+
+    /* we need FI_EP_RDM */
+    if (info->ep_attr->type != FI_EP_RDM)
+        return OPAL_ERROR;
+
+    mr_mode = info->domain_attr->mr_mode;
+
+    if (!(mr_mode == FI_MR_BASIC || mr_mode == FI_MR_SCALABLE ||
+         (mr_mode & ~(FI_MR_VIRT_ADDR | FI_MR_ALLOCATED | FI_MR_PROV_KEY)) == 0))
+        return OPAL_ERROR;
+    }
+
+    /* provider specifics */
+    if (!strcmp(prov_name, "sockets")) {
+        return socks_validate_info(info);
+    } else if (!strcmp(prov_name, "psm2")) {
+        return psm2_validate_info(info);
+    } else if (!strcmp(prov_name, "gni")) {
+        return gni_validate_info(info);
```

Comment:
Why are you hard-coding provider names / logic?

Reply:
I don't like it neither but what would be a better way to do it? I took a look at MTL/OFI and they are more or less the same by converting string into some enum. So I think the hard-coding has to be there somewhere right?

Reply:
Shouldn't you validate (and set hints) based on what the BTL needs?

Reply:
Yeah, right now those validate info functions are just empty. It was in early development where we have to force certain flags for certain providers. 

I still keep it because some provider might really need it. i.e,  provider returns FI_MR_UNSPEC and we want to force only FI_MR_VIRT_ADDR.

Reply:
some provider also return FI_MR_UNSPEC which in API 1.5 assume FI_MR_SCALABLE but in some case we want FI_MR_BASIC (which provider totally support) so I have a place for the providers/dev to force it themselves.

Reply:
I see what you are saying here. Let's not have this provider specific thing. 
will fix in upcoming commit.